.github/workflows/build-and-test.yml

```diff
@@ -215,6 +215,171 @@ jobs:
         if: ${{ failure() && runner.debug }}
         uses: canonical/action-tmate@main
 
+  juju-cluster-maintenance-test:
+    needs:
+      - lint
+      - unit-test
+      - build
+    name: Juju cluster maintenance test
+    runs-on: [self-hosted, xlarge]
+    steps:
+
+      - name: Download charm
+        uses: actions/download-artifact@v3
+        with:
+          name: charms
+          path: ~/artifacts/
+
+      - name: Checkout code
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+
+      - name: Setup LXD
+        uses: canonical/setup-lxd@v0.1.1
+        with:
+          # pin lxd to LTS release.
+          channel: 5.21/stable
+
+      - name: Install dependencies
+        run: ./tests/scripts/ci_helpers.sh install_deps
+
+      - name: Install MicroCeph charm
+        run: |
+          date
+          juju deploy -n 4 ~/artifacts/microceph.charm --config default-pool-size=1
+          ./tests/scripts/ci_helpers.sh wait_for_microceph_bootstrap
+          date
+
+      - name: Show Juju status
+        run: |
+          set -eux
+          juju status
+
+      - name: Enable one extra ceph mon for redundancy
+        run: |
+          set -eux
+          juju ssh microceph/3 -- sudo microceph enable mon
+
+      - name: Add disks
+        run: |
+          set -eux
+          juju run microceph/0 add-osd loop-spec=4G,1 --wait=2m
+          juju run microceph/1 add-osd loop-spec=4G,1 --wait=2m
+          juju run microceph/2 add-osd loop-spec=4G,1 --wait=2m
+          juju run microceph/3 add-osd loop-spec=4G,1 --wait=2m
+
+      - name: Show Ceph status
+        run: |
+          set -eux
+          juju ssh microceph/0 -- sudo microceph status
+          juju ssh microceph/0 -- sudo microceph.ceph -s
+
+      - name: Dry run test
+        run: |
+          set -eux
+
+          action_counts=$(juju run microceph/0 exit-maintenance dry-run=true --format json | jq '."microceph/0".results.actions | length')
+          if [ "$action_counts" != "3" ] ; then
+            echo "Expect dry run exit-mainteance to produce 3 actions."
```

Comment:
can the expectation of the test case be mentioned explicitly as a code comment ?

Reply:
also, nit-picking but should use similar messages for failures, someplaces it expects `3 actions` but other places it says `3 action plan`.

Reply:
fixed, `steps` is probably better

Reply:
ack, thanks